imageio/plugins/__init__.py

```diff
@@ -79,40 +79,6 @@
 
 """
 
-from ..core.imopen import imopen
-
-try:
-    import PIL
-    from .pillow import PillowPlugin
-
-    imopen.register_plugin("pillow", PillowPlugin)
-except ImportError:
-    pass  # Pillow not installed
-
-
-# First import plugins that we want to take precedence over freeimage
-from . import tifffile
-from . import fits  # depends on astropy
-from . import pillow_legacy
-from . import grab
-
-from . import freeimage
-from . import freeimagemulti
-
-from . import ffmpeg
-
-from . import bsdf
-from . import dicom
-from . import npz
-from . import swf
-
-from . import feisem  # special kind of tiff, uses _tiffile
-from . import simpleitk  # depends on itk or SimpleITK
-from . import gdal  # depends on gdal
-
-from . import lytro
-from . import spe
```

Comment:
Does all this mean that a user would have to `import imagio.plugins.pillow` in order to use the Pillow plugin?

Reply:
Nope 🚀 We run this for the user ([link to relevant lines](https://github.com/FirefoxMetzger/imageio/blob/68819d27a669364daba53e271e4583868c81b335/imageio/config/plugins.py#L101-L102)).

Roughly, what happens now is 

```python

import imageio as iio 

# modules imported up to this point
# - imageio
# - numpy

iio.imread("some.tiff")  # tifffile is automatically imported here

# modules imported up to this point
# - imageio
# - numpy
# - tifffile

iio.imread("imageio:chelsea.png")  # pillow is automatically imported here

# modules imported up to this point
# - imageio
# - numpy
# - tifffile
# - pillow

iio.imread("some.bogus")  # automatically imports and tries all plugins listed in `iio.config.known_plugins`
# raises IOError

# modules imported up to this point
# all of them ... we had to load them to try our best to read `some.bogus`
```

The above introduces a one-time cost for each plugin's first `imread`, but that can be avoided by your suggestion, i.e., if the user manually calls `import imagio.plugins.pillow` then the one-time cost for loading the plugin is paid there instead of being paid during `imread`.

On the flip side, if a user never uses `imageio-ffmpeg`, then he/she will never pay the cost of importing it (nor will he/she ever face exceptions because that plugin doesn't work on their machine). It also avoids us doing the "is this backend installed" dance inside a plugin. The plugin simply assumes that it is installed, and if it isn't it fails with an ImportError, which `imopen` picks up and handles appropriately.

Reply:
I agree that this approach has certain advantages. I think I also prefer it over the earlier approach.

One note though, is that we may want to take measures for people freezing apps with imageio. See e.g. https://github.com/pygfx/wgpu-py/tree/main/wgpu/__pyinstaller

Reply:
@almarklein I've been meaning to ask about the files related to freezing support in our repo. I never come across this before and haven't seen it anywhere else. When I googled it a while back I only found old resources that seemed to be from the python2.7 era, and I couldn't find anything in the official python docs (probably not looking at the right place?)

Is this still a thing that people actually use? And, consequentially, is it worth supporting? What would it be used for?

Reply:
Yes, people definitely freeze apps. E.g to create apps to distribute to non-coders. Since a year or so, PyInstaller provided an API that packages can use to tell pyinstaller what extra data and modules must be included. See these docs: https://pyinstaller.readthedocs.io/en/stable/hooks.html#provide-hooks-with-package, and also the links in my post above (I added a 2nd link).

Reply:
Okay, then we keep supporting freezing. I am not an expert here, but I doubt we need to do much extra work. The burden of including all required modules lies with the packaging utility, and the plugin code itself is still part of the repo. We simply delay loading it for as long as possible.

Reply:
Tools like pyinstaller and cx_freeze look at the actual imports at the top of the file, so it probably won't include the plugins now :)  But implementing the hooks should be relatively simple. Its also easy to test using PyInstaller tooling. I could have a look at this soonish.

Reply:
> Tools like pyinstaller and cx_freeze look at the actual imports at the top of the file,

This sounds quite scary. Grouping imports at the top of the file is usually fine, but dynamic imports (e.g. with a `import` nested inside a function or via `importlib`) have valid use-cases. `importlib` is part of the python spec since v3.1 so I would hope that packaging tools have adjusted to that by now...

Anyways, let's figure out how this works and then take the appropriate actions from there. The solution can be as simple as adding a note to the docs that says "If you want to freeze your script, remember to import all the plugins your script need at the top".

Reply:
> This sounds quite scary.

Most freeze apps do it this way, and there are very valid reasons for it. If an import is not in the module level, you cannot know whether it will be imported at runtime. If it would be included regardless, it would add up to the size of the final binary (especially if those imports pull in other imports etc). Also, these imports may well be platform specific.

Tracking this in #684